Add an example of starting Exshome

RomanKotov opened the pull request:
This example shows how to start Elixir smart home (Exshome) from a single script.

ref: https://github.com/exshome/exshome

RomanKotov committed:
Add an example of starting Exshome

RomanKotov committed:
Allow IEx to attach to the script

RomanKotov commented:
Have replaced `Process.sleep(:infinity)` with `System.no_halt(true)`. It allows to debug a script via IEx shell, like `iex exshome.exs`. Iex shell was not able to attach without it.

wojtekmach commented:
On the first run I'm getting:

```
13:51:34.621 [notice] Application exshome exited: Exshome.Application.start(:normal, []) returned an error: shutdown: failed to start child: Exshome.Dependency.GenServerDependency.DependencySupervisor
    ** (EXIT) shutdown: failed to start child: ExshomeClock.Services.LocalTime
        ** (EXIT) an exception was raised:
            ** (Exqlite.Error) no such table: service_settings
SELECT s0."name", s0."data", s0."version", s0."inserted_at", s0."updated_at" FROM "service_settings" AS s0 WHERE (s0."name" = ?)
                (ecto_sql 3.8.3) lib/ecto/adapters/sql.ex:932: Ecto.Adapters.SQL.raise_sql_call_error/1
                (ecto_sql 3.8.3) lib/ecto/adapters/sql.ex:847: Ecto.Adapters.SQL.execute/6
                (ecto 3.8.4) lib/ecto/repo/queryable.ex:221: Ecto.Repo.Queryable.execute/4
                (ecto 3.8.4) lib/ecto/repo/queryable.ex:19: Ecto.Repo.Queryable.all/3
                (ecto 3.8.4) lib/ecto/repo/queryable.ex:147: Ecto.Repo.Queryable.one/3
                (exshome 0.1.4) lib/exshome/settings/schema.ex:27: Exshome.Settings.Schema.get_or_create/2
                (exshome 0.1.4) lib/exshome/settings.ex:14: Exshome.Settings.get_settings/1
                (exshome 0.1.4) lib/exshome/dependency.ex:20: Exshome.Dependency.subscribe/1
```

is this expected?

wojtekmach commented on exshome.exs:
```diff
@@ -0,0 +1,35 @@
+root_folder = "HOME" |> System.get_env() |> Path.join(".exshome")
```
please change this to `tmp/exshome`. I'd rather not have these scripts have any footprint but if they need to, keep it in this dir.

RomanKotov replied:
Thanks for a suggestion, I have updated it.

wojtekmach commented on exshome.exs:
```diff
@@ -0,0 +1,35 @@
+
+Exshome.Release.migrate()
+Application.ensure_all_started(:exshome)
+System.no_halt(true)
```
nice 👍

wojtekmach commented:
<img width="454" alt="image" src="https://user-images.githubusercontent.com/76071/175292745-f74e28f4-a1b8-4a35-a546-89e94a4f28b2.png">

are these missing dependencies expected?

RomanKotov committed:
Use temporary folder as Exshome root

RomanKotov commented:
> On the first run I'm getting:
> 
> ```
> 13:51:34.621 [notice] Application exshome exited: Exshome.Application.start(:normal, []) returned an error: shutdown: failed to start child: Exshome.Dependency.GenServerDependency.DependencySupervisor
>     ** (EXIT) shutdown: failed to start child: ExshomeClock.Services.LocalTime
>         ** (EXIT) an exception was raised:
>             ** (Exqlite.Error) no such table: service_settings
> SELECT s0."name", s0."data", s0."version", s0."inserted_at", s0."updated_at" FROM "service_settings" AS s0 WHERE (s0."name" = ?)
>                 (ecto_sql 3.8.3) lib/ecto/adapters/sql.ex:932: Ecto.Adapters.SQL.raise_sql_call_error/1
>                 (ecto_sql 3.8.3) lib/ecto/adapters/sql.ex:847: Ecto.Adapters.SQL.execute/6
>                 (ecto 3.8.4) lib/ecto/repo/queryable.ex:221: Ecto.Repo.Queryable.execute/4
>                 (ecto 3.8.4) lib/ecto/repo/queryable.ex:19: Ecto.Repo.Queryable.all/3
>                 (ecto 3.8.4) lib/ecto/repo/queryable.ex:147: Ecto.Repo.Queryable.one/3
>                 (exshome 0.1.4) lib/exshome/settings/schema.ex:27: Exshome.Settings.Schema.get_or_create/2
>                 (exshome 0.1.4) lib/exshome/settings.ex:14: Exshome.Settings.get_settings/1
>                 (exshome 0.1.4) lib/exshome/dependency.ex:20: Exshome.Dependency.subscribe/1
> ```
> 
> is this expected?

There is no sqlite3 database when the application is started the first time.

`Mix.install/2` part installs the library and also starts an `:exshome` application. It seems, that there is no option to skip starting it. Possibly it would be nice to only install the libraries, but not start the applications.

`Exshome.Release.migrate()` part runs the migrations, and creates a database.

I have added `Application.ensure_all_started(:exshome)` line just to make sure, that application is started even if `Mix.install/2` will not start it.

So, I think, this error (or similar) may appear when there is no database, or migrations are not completed yet.

I am sure, that errors in logs are not normal, but this one is expected.

I understand, I need to wait until the database and all migrations are ready and then start an application, to remove this error.

I see the following constraints:
- `Mix.install/2` can run only once per script, so I will not be able to install ecto-related dependencies.
- I will not be able to delay starting the application (have tried adding `runtime: false` or `app: false` to the dependency).

I could delay starting the application itself, until the database is ready. It sounds like a supervisor that starts its children only when some requirements are met.

Please, can you advise how to fix it properly?

RomanKotov commented:
> <img alt="image" width="454" src="https://user-images.githubusercontent.com/76071/175292745-f74e28f4-a1b8-4a35-a546-89e94a4f28b2.png">
> 
> are these missing dependencies expected?

Yes, the "player" part of application requires [mpv](https://mpv.io) to work properly.
Each page shows "Missing dependencies" placeholder if some of its dependencies are not started yet, or not available.

Should I add a comment about `MPV` requirement to this PR?

RomanKotov committed:
Update Exshome version

RomanKotov commented:
Thank you for your feedback.
I have updated the PR with fixes.
1. Migration now runs before starting main application, so it should be no `[notice] Application exshome exited` errors.
2. I have removed "player" application from this script. So it uses only clock now and does not show any "Missing dependencies".

wojtekmach commented:
Great, thank you!

RomanKotov commented:
Thank you!
